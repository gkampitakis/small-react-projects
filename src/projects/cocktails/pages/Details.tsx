import React, { ReactElement, useState, useEffect } from 'react';
import { useParams, Link } from 'react-router-dom';
import { useFetch } from '../../../hooks';
import ErrorCmp from '../../../components/Error';
import Loading from '../../../components/Loading';


const URL = 'https://www.thecocktaildb.com/api/json/v1/1/lookup.php?i=';

interface CocktailDetails {
  ingredients: string[];
  name: string;
  image: string;
  info: string;
  category: string;
  glass: string;
  instructions: string;
}

export default function Details (): ReactElement {
  const { id } = useParams<{ id: string }>();
  const [cocktail, setCocktail] = useState<CocktailDetails>({} as any);
  const [data, loading, _, error] = useFetch<any>(`${URL}${id}`);

  function extractDetails (data: any): CocktailDetails {
    const {
      strDrink,
      strDrinkThumb,
      strAlcoholic,
      strCategory,
      strGlass,
      strInstructions,
    } = data.drinks[0];
    const ingredients: string[] = [];

    for (let i = 1; i < 16; i++) {
      const ingredient = data.drinks[0][`strIngredient${i}`];
      if (!ingredient) break;
      ingredients.push(ingredient);
    }

    return {
      name: strDrink,
      image: strDrinkThumb,
      info: strAlcoholic,
      category: strCategory,
      glass: strGlass,
      instructions: strInstructions,
      ingredients
    };
  }

  useEffect(() => {
    if (data?.drinks[0]) {
      setCocktail(extractDetails(data));
    }
  }, [data]);

  if (loading) {
    return <Loading />;
  }

  if (error) {
    return <ErrorCmp message={'Error occurred'} />;
  }


  return (
    <section className="cocktail-section">
      <Link to="/cocktails" className="btn">Back Home</Link>
      {!cocktail.name && <h2>Cocktail details not found</h2>}
      {cocktail.name && <>
        <h2 className="title">{cocktail.name}</h2>
        <div className="drink">
          <img src={cocktail.image} alt={cocktail.image} />
          <div className="details">
            <p>
              <span className='drink-data'>name :</span> {cocktail.name}
            </p>
            <p>
              <span className='drink-data'>category :</span> {cocktail.category}
            </p>
            <p>
              <span className='drink-data'>info :</span> {cocktail.info}
            </p>
            <p>
              <span className='drink-data'>glass :</span> {cocktail.glass}
            </p>
            <p>
              <span className='drink-data'>instructions :</span> {cocktail.instructions}
            </p>
            <p>
              <span className='drink-data'>ingredients :</span>
              {cocktail?.ingredients?.map((item, index) => <span key={index}> {item}</span>)}
            </p>
          </div>
        </div>
      </>
      }
    </section>
  );
}
