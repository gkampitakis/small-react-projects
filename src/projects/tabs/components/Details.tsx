import React, { ReactElement } from 'react';
import { FaAngleDoubleRight } from 'react-icons/fa'


interface DetailsProps {
  data: {
    id: string;
    order: number;
    title: string;
    dates: string;
    duties: string[];
    company: string;
  }
}


export default function Details ({ data }: DetailsProps): ReactElement {
  return (
    <section className="details">
      <header>
        <h3>{data.title}</h3>
        <h4>{data.company}</h4>
        <p className="date">{data.dates}</p>
      </header>
      {
        data.duties.map((d, idx) => (
          <div className="job-desc" key={idx}>
            <FaAngleDoubleRight className="duty-icon" />
            <p>{d}</p>
          </div>
        ))
      }
    </section>
  );
}
