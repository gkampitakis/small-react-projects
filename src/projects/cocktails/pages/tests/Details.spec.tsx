import React from 'react';
import { render, act, screen, waitForElementToBeRemoved } from '@testing-library/react';
import Details from '../Details';
import { MemoryRouter } from 'react-router-dom';
import { FetchMock } from 'jest-fetch-mock/types';


const fetchMock = (fetch as FetchMock);
const mockData = {
  drinks: [
    {
      strDrink: 'mockDrink',
      strDrinkThumb: 'mockThumb',
      strAlcoholic: 'mockInfo',
      strCategory: 'mockCategory',
      strGlass: 'mockGlass',
      strInstructions: 'mockInstructions',
      strIngredient1: 'mockIngredient1',
      strIngredient2: 'mockIngredient2',
      strIngredient3: 'mockIngredient3',
      strIngredient4: 'mockIngredient4',
    }
  ]
};

describe('Details', () => {
  beforeAll(fetchMock.enableMocks);
  beforeEach(fetchMock.resetMocks);

  it('Should load correctly', async () => {
    fetchMock.mockResponseOnce(JSON.stringify(mockData));
    await act(async () => {
      const { baseElement } = render(<Details />,
        { wrapper: MemoryRouter });

      expect(baseElement).toMatchSnapshot();
    });
  });

  it('Should load data after loading', async () => {
    fetchMock.mockResponseOnce(JSON.stringify(mockData));
    const { baseElement } = render(<Details />,
      { wrapper: MemoryRouter });

    await waitForElementToBeRemoved(screen.getByAltText(/Loading/i));

    expect(baseElement).toMatchSnapshot();
  });

  it('Should show message if no cocktails found', async () => {
    fetchMock.mockResponseOnce(JSON.stringify({ drinks: [] }));
    const { baseElement } = render(<Details />,
      { wrapper: MemoryRouter });

    await waitForElementToBeRemoved(screen.getByAltText(/Loading/i));

    expect(baseElement).toMatchSnapshot();
  });

  it('Should show error', async () => {
    fetchMock.mockRejectOnce(() => Promise.reject('Mock Error'));

    const { baseElement } = render(<Details />,
      { wrapper: MemoryRouter });

    await waitForElementToBeRemoved(screen.getByAltText(/Loading/i));

    expect(baseElement).toMatchSnapshot();
  });
});
