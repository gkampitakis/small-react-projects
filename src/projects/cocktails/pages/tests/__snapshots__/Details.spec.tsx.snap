// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Details Should load correctly 1`] = `
<body>
  <div>
    <main
      class="loading_component"
    >
      <div
        class="loader"
      >
        <div
          class="bar"
        />
      </div>
      <img
        alt="Loading Logo"
        class="logo"
        src="loading.png"
      />
    </main>
  </div>
</body>
`;

exports[`Details Should load data after loading 1`] = `
<body>
  <div>
    <section
      class="cocktail-section"
    >
      <a
        class="btn"
        href="/cocktails"
      >
        Back Home
      </a>
      <h2
        class="title"
      >
        mockDrink
      </h2>
      <div
        class="drink"
      >
        <img
          alt="mockThumb"
          src="mockThumb"
        />
        <div
          class="details"
        >
          <p>
            <span
              class="drink-data"
            >
              name :
            </span>
             
            mockDrink
          </p>
          <p>
            <span
              class="drink-data"
            >
              category :
            </span>
             
            mockCategory
          </p>
          <p>
            <span
              class="drink-data"
            >
              info :
            </span>
             
            mockInfo
          </p>
          <p>
            <span
              class="drink-data"
            >
              glass :
            </span>
             
            mockGlass
          </p>
          <p>
            <span
              class="drink-data"
            >
              instructions :
            </span>
             
            mockInstructions
          </p>
          <p>
            <span
              class="drink-data"
            >
              ingredients :
            </span>
            <span>
               
              mockIngredient1
            </span>
            <span>
               
              mockIngredient2
            </span>
            <span>
               
              mockIngredient3
            </span>
            <span>
               
              mockIngredient4
            </span>
          </p>
        </div>
      </div>
    </section>
  </div>
</body>
`;

exports[`Details Should show error 1`] = `
<body>
  <div>
    <main
      class="error_component"
    >
      <img
        alt="Error page image"
        src="error.svg"
      />
      <footer>
        <h2>
          Error occurred
        </h2>
      </footer>
    </main>
  </div>
</body>
`;

exports[`Details Should show message if no cocktails found 1`] = `
<body>
  <div>
    <section
      class="cocktail-section"
    >
      <a
        class="btn"
        href="/cocktails"
      >
        Back Home
      </a>
      <h2>
        Cocktail details not found
      </h2>
    </section>
  </div>
</body>
`;
